File: old/eden2d/ts/Message.ts
module eden {

  export const MessageTypeConnect      = 0;
  export const MessageTypeConnected    = 1;
  export const MessageTypeRegionUpdate = 2;
  export const MessageTypePlayerAction = 3;

  export const PlayerActionMove    = 0; // [MoveDir]
  export const PlayerActionPlace   = 1; // [ObjectType, Position]

  export const MoveNorth = 1;
  export const MoveSouth = 2;
  export const MoveWest  = 3;
  export const MoveEast  = 4;

  export interface Message {
    Type: number;
    Connect: MessageConnect;
    Connected: MessageConnected;
    RegionUpdate: MessageRegionUpdate;
    PlayerAction: MessagePlayerAction;
  }

  export interface MessageConnect {
    Name: string;
    Region: string;
  }

  export interface MessageConnected {
    Width: number;
    Height: number;
    Cells: number[];
    Objects: ObjectRecord[];
  }

  export interface TileRecord {
    I: number; // y * width + x + cell
    V: number; // new value
  }

  export interface ObjectRecord {
    I: number; // Id
    T: number; // Type -- 0 == remove
    P: number; // Pos
    N: number; // Next
  }

  export interface MessageRegionUpdate {
    TileRecs: TileRecord[];
    ObjectRecs: ObjectRecord[];
  }

  export interface MessagePlayerAction {
    Type: number;
    Args: any[];
  }
}

File: old/eden2d/go/src/eden/proto/message.go
package proto

type MessageType uint16

const (
	MessageTypeConnect      MessageType = 0
	MessageTypeConnected                = 1
	MessageTypeRegionUpdate             = 2
	MessageTypePlayerAction             = 3
)

type PlayerActionType uint8

const (
	PlayerActionMove  PlayerActionType = 0 // [MoveDir]
	PlayerActionPlace                  = 1 // [ObjectType, Position]
)

type MoveDir uint8

const (
	MoveNorth MoveDir = 1
	MoveSouth         = 2
	MoveWest          = 3
	MoveEast          = 4
)

type Message struct {
	Type         MessageType
	Connect      *MessageConnect      `json:",omitempty"`
	Connected    *MessageConnected    `json:",omitempty"`
	RegionUpdate *MessageRegionUpdate `json:",omitempty"`
	PlayerAction *MessagePlayerAction `json:",omitempty"`
}

type MessageConnect struct {
	Name   string
	Region string
}

type MessageConnected struct {
	Width, Height uint16
	Cells         []uint32
	Objects       []ObjectRecord
}

type ObjectRecord struct {
	I uint32 // Id
	T uint32 // Type -- 0 == remove
	P uint8  // Pos
	N uint32 // Next
}

type TileRecord struct {
	I uint32 // y * width + x + layer
	V uint32 // new value
}

type MessageRegionUpdate struct {
	TileRecs   []TileRecord
	ObjectRecs []ObjectRecord
}

type MessagePlayerAction struct {
	Type PlayerActionType
	Args []interface{}
}
